cherry-rewrite: storage: sprinkle log events through push txn queue

tbg opened the pull request:
Motivated by reports of a user about transactions getting stuck
in the PushTxnQueue which will be easier to diagnose with this
additional information.

Had to rewrite this a bunch due to the enhancements that have landed
on master but not here. Consider this new code.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16741)
<!-- Reviewable:end -->

tbg committed:
cherry-rewrite: storage: sprinkle log events through push txn queue

Motivated by reports of a user about transactions getting stuck
in the PushTxnQueue which will be easier to diagnose with this
additional information.

Had to rewrite this a bunch due to the enhancements that have landed
on master but not here. Consider this new code.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16741#-:-KnkYIcqci0hSYFpZmIb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
